### PR TITLE
Fix failure for scalar subquery in value-side of quantified comparison

### PR DESCRIPTION
**Pull request by martint**

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/SubqueryPlanner.java [hunk 1]
```diff
@@ -117,7 +117,7 @@ public PlanBuilder handleSubqueries(PlanBuilder builder, Expression expression,
         for (Cluster<InPredicate> cluster : cluster(builder.getScope(), selectSubqueries(builder, expression, subqueries.getInPredicatesSubqueries()))) {
             builder = planInPredicate(builder, cluster, subqueries);
         }
-        for (Cluster<SubqueryExpression> cluster : cluster(builder.getScope(), selectSubqueries(builder, expression, subqueries.getScalarSubqueries()))) {
+        for (Cluster<SubqueryExpression> cluster : cluster(builder.getScope(), selectSubqueries(builder, expression, subqueries.getSubqueries()))) {
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/SubqueryPlanner.java:120]: The rename in this and other classes could be extracted to another commit.
In fact, it logically belongs to this PR: https://github.com/trinodb/trino/pull/7774

File: core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java [hunk 2]
```diff
@@ -1301,4 +1301,12 @@ public void testCorrelatedUnnestInLateralSubquery()
                         "ON TRUE "))
                 .matches("VALUES 1, 2, 3, 4, 5, null");
     }
+
+    @Test
+    public void testQuantifiedComparison()
```
Comment (kasiafi) [inline core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java:1306]: Is there a similar test with in-predicate? If no, could you add one?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -1903,11 +1900,13 @@ protected Type visitExists(ExistsPredicate node, StackableAstVisitorContext<Cont
...
         protected Type visitQuantifiedComparisonExpression(QuantifiedComparisonExpression node, StackableAstVisitorContext<Context> context)
         {
+            quantifiedComparisons.add(NodeRef.of(node));
+
             Expression value = node.getValue();
             process(value, context);
 
             Expression subquery = node.getSubquery();
-            process(subquery, context);
+            setExpressionType(subquery, analyzeSubquery((SubqueryExpression) subquery, context));
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1909]: `analyzeSubquery` is always accompanied by `setExpressionType`. Could `analyzeSubquery` do 
```
return(setExpressionType(...))
```
instead?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 4]
```diff
@@ -1810,16 +1810,17 @@ protected Type visitInPredicate(InPredicate node, StackableAstVisitorContext<Con
...
 
                 coerceToSingleType(context,
                         "IN value and list items must be the same type: %s",
                         ImmutableList.<Expression>builder().add(value).addAll(inListExpression.getValues()).build());
             }
             else if (valueList instanceof SubqueryExpression) {
+                subqueryInPredicates.add(NodeRef.of(node));
+                setExpressionType(valueList, analyzeSubquery((SubqueryExpression) valueList, context));
                 coerceToSingleType(context, node, "value and result of subquery must be of the same type for IN expression: %s vs %s", value, valueList);
             }
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1825]: else throw?

**Review by kasiafi — COMMENTED** [review]
It looks super clean now. Thanks.

File: core/trino-main/src/main/java/io/trino/sql/planner/SubqueryPlanner.java (hunk 1 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/SubqueryPlanner.java:120]: Pulled it out into a separate commit.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java (hunk 3 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java]: Yes, that's leftover from a previous experiment.

File: core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java (hunk 2 above)
Comment (martint) [reply core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java:1306]: Added

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java (hunk 4 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1825]: Yes, but that's unrelated to this PR. I'll add a separate commit.

**Comment by kasiafi** [comment]
Looks good to me.

**Review by dain — APPROVED** [review]
